SOD.Common/Helpers/PlayerStatus.cs

```diff
@@ -0,0 +1,181 @@
+            }
+            if (force && IllegalStatusModifierDictionary.ContainsKey(key))
+            {
+                // Remove entry with the same key, if it exists. Do not stop
+                // the coroutine if it is started.
+                IllegalStatusModifierDictionary.Remove(key);
+            }
+            else if (!force)
+            {
+                // Modifier with this key is already present
+                return;
+            }
+            IllegalStatusModifierDictionary.Add(key, new IllegalStatusModifier(key, duration, duration == 0.0f ? false : true));
+            if (UpdateIllegalStatusModifiersCoroutine == null)
+            {
+                UpdateIllegalStatusModifiersCoroutine = RuntimeHelper.StartCoroutine(UpdateIllegalStatusModifiers());
+            }
+        }
+
+        /// <summary>
+        /// Removes an illegal status modifier, if present.
+        /// </summary>
+        /// <param name="key">The key identifying the modifier.</param>
+        public void RemoveIllegalStatusModifier(string key)
+        {
+            if (key == "")
+            {
+                // No need to throw
+                return;
+            }
+            if (IllegalStatusModifierDictionary.ContainsKey(key))
+            {
+                IllegalStatusModifierDictionary.Remove(key);
+            }
+            if (IllegalStatusModifierDictionary.Count > 0)
+            {
+                return;
+            }
+            if (UpdateIllegalStatusModifiersCoroutine != null)
+            {
+                RuntimeHelper.StopCoroutine(UpdateIllegalStatusModifiersCoroutine);
+                UpdateIllegalStatusModifiersCoroutine = null;
+            }
+        }
+
+        /// <summary>
+        /// Toggles an illegal status modifier by adding it if the key is not present, or removing it if the key is present. If added, the modifier will not have a duration associated with it (it will be active indefinitely, until toggled or removed).
+        /// </summary>
+        /// <param name="key">The key identifying the modifier.</param>
+        public void ToggleIllegalStatusModifier(string key)
+        {
+            if (IllegalStatusModifierDictionary.ContainsKey(key))
+            {
+                RemoveIllegalStatusModifier(key);
+                return;
+            }
+            SetIllegalStatusModifier(key);
+        }
+
+        internal IEnumerator UpdateIllegalStatusModifiers()
+        {
+            var setAsIllegal = false;
+            var modifierKeysToRemove = new List<string>();
+            var deltaTime = 0.0f;
+            while (true)
+            {
+                yield return new WaitForEndOfFrame();
+
+                if (Lib.Time.IsGamePaused)
+                {
+                    continue;
+                }
+
+                setAsIllegal = false;
+                deltaTime = UnityEngine.Time.deltaTime;
+
+                foreach (var key in IllegalStatusModifierDictionary.Keys)
```

Comment:
I'm just wondering, if this coroutine is running in this foreach while I add a new modifier will it not give the "cannot modify while enumerating collection" error?

Reply:
When is it actually ever breaking out of this while loop from the coroutine itself? never?